File: Subarrays with equal 1s and 0s.java
class Solution {
    public int countSubarray(int[] arr) {
        HashMap<Integer, Integer> hm = new HashMap<>();
        int ans = 0;
        int sum = 0;
        hm.put(0, 1);
        for(int num : arr){
            if(num == 0) sum--;
            else sum++;
            
            if(hm.containsKey(sum)){
                ans += hm.get(sum);
                hm.put(sum, hm.get(sum) + 1);
            }
            else{
                hm.put(sum, 1);
            }
        }
        
        return ans;
    }
}

File: Subarrays with equal 1s and 0s.cpp
class Solution {
public:
    int countSubarray(const vector<int>& arr) {
        unordered_map<int, int> hm;
        int ans = 0;
        int sum = 0;
        hm[0] = 1;
        for (int num : arr) {
            if (num == 0) sum--;
            else sum++;

            if (hm.find(sum) != hm.end()) {
                ans += hm[sum];
                hm[sum]++;
            } else {
                hm[sum] = 1;
            }
        }

        return ans;
    }
};